ops/charm.py

```diff
@@ -9,17 +9,27 @@ class HookEvent(EventBase):
 
 class ActionEvent(EventBase):
 
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._params = None
+
     def defer(self):
         raise RuntimeError('cannot defer action events')
 
     def restore(self, snapshot):
+        self._params = None
```

Comment:
why do we need to reset _params to None here.
Are params not part of the saved snapshot?

Reply:
This is just a side-effect of how we emit events: they are persisted and reemitted. This results in a  strange behavior for `ActionEvents` because it happens for them as well while they cannot be manually deferred.

The problem with restoring event objects is that an `__init__` method is not called for unpickled objects and the framework calls `restore` on them:
https://github.com/canonical/operator/blob/ed093d5c3fa16454f87639550464376446e1713e/ops/framework.py#L490-L494

I don't think we need to store params into a unit-local state snapshot just to read them back right away. Similarly, we do not do anything like that with config on `config-changed` or relation data for relation data events.